Feature/lamdera format

CharlonTank opened the pull request:
https://github.com/orgs/lamdera/projects/2/views/1?pane=issue&itemId=90905452
This pull request introduces a new feature to format Elm source files using the `lamdera format` command. The most important changes include adding the new `Lamdera.CLI.Format` module, updating existing modules to support the new command, and modifying the `Lamdera.CLI` module to include the new format command.

### New Feature: Elm Source File Formatting

* Added `Lamdera.CLI.Format` module to implement the `lamdera format` command, which includes functionality to format Elm source files, handle various command-line options, and provide usage instructions.
* Updated `Lamdera.CLI` module to export the new `format` function and include the format command in the CLI interface. [[1]](diffhunk://#diff-cb80f51d7a042d7b8fc239cb6c9ca8c91a35553fbee442227815d8195fb7f814L3-R3) [[2]](diffhunk://#diff-cb80f51d7a042d7b8fc239cb6c9ca8c91a35553fbee442227815d8195fb7f814R19) [[3]](diffhunk://#diff-cb80f51d7a042d7b8fc239cb6c9ca8c91a35553fbee442227815d8195fb7f814R197-R223)
* Modified `elm.cabal` to include the new `Lamdera.CLI.Format` module.

### Codebase Enhancements

* Updated `Ext.ElmFormat` module to support file path input for formatting functions and handle new imports required for formatting Elm source files. [[1]](diffhunk://#diff-85c281d479ab2d11635d5dd622b188cf2c25b0c3142043303a7c065c892bfef3R12) [[2]](diffhunk://#diff-85c281d479ab2d11635d5dd622b188cf2c25b0c3142043303a7c065c892bfef3R24-R52)
* Added the new `format` command to the list of available commands in the `main` function of `terminal/src/Main.hs`.

CharlonTank committed:
Improve format command to match elm-format behavior

CharlonTank committed:
Improve format command to match elm-format behavior

CharlonTank committed:
Improve elm format error messages and file handling

CharlonTank committed:
Match elm-format behavior for file formatting confirmation

supermario commented on ext-common/Ext/ElmFormat.hs:
```diff
@@ -20,31 +21,35 @@ import qualified ElmFormat.Cli
 formatOrPassthrough :: Text -> Text
 formatOrPassthrough text = do
-  case format text of
+  case format "stdin" text of
```
Should this be `"stdin:nofilepath"` to match the previous mode?

https://github.com/lamdera/compiler/pull/39/files#diff-85c281d479ab2d11635d5dd622b188cf2c25b0c3142043303a7c065c892bfef3L27

supermario commented on extra/Lamdera/CLI/Format.hs:
```diff
@@ -0,0 +1,245 @@
+      TIO.putStrLn "  --yes                 Reply 'yes' to all automated prompts."
+      TIO.putStrLn "  --validate            Check if files are formatted without changing them."
+      TIO.putStrLn "  --stdin               Read from stdin, output to stdout.\n"
+      TIO.putStrLn "Note: All flags must use double dashes (--flag). Single dashes (-flag) are not supported.\n"
```
Probably don't need this line? One less difference to elm-format.

supermario commented on extra/Lamdera/CLI/Format.hs:
```diff
@@ -0,0 +1,245 @@
+      TIO.putStrLn "  lamdera format src/                         # format all *.elm files in the src directory\n"
+      TIO.putStrLn "Full guide to using elm-format at <https://github.com/avh4/elm-format>"
+      Exit.exitSuccess
+    else case inputs of
```
Wondering if it was possible to patch directly into elm-format's existing handling code here rather than replicating it? Or was there some problem doing so?

supermario reviewed (commented):
Tested locally, looking good! A few questions.

CharlonTank committed:
Address PR feedback: use stdin:nofilepath and remove redundant note

CharlonTank commented:
Thanks for the review @supermario\! I apologize for the delayed response - it's been several months and I've had a lot on my plate.

I've investigated patching directly into elm-format's CLI handling code as you suggested.

While it would be ideal to reuse elm-format's CLI code directly, I found that the two CLIs have different architectures:
- Lamdera uses the `Terminal` module with its own command parsing structure (consistent across all Lamdera commands)
- elm-format uses a different command-line parsing approach with its own `CommandLine.Program` monad

Directly integrating would require either:
1. Special-casing the format command to bypass Lamdera's normal CLI flow, or
2. Significant refactoring of how Lamdera handles commands

The current implementation:
- Maintains consistency with other Lamdera commands (same help format, flag handling, etc.)
- Still uses elm-format's core formatting logic via the `ElmFormat.Cli.format` API
- Follows the Lamdera project's minimal-invasion philosophy
- Makes it easier to maintain when elm-format updates (we only depend on the stable `format` function API)

That said, if you feel strongly about deeper integration, I'm happy to explore it further. The trade-off would be more coupling with elm-format's internals vs. the current approach of depending only on its public API.

supermario commented:
> Special-casing the format command to bypass Lamdera's normal CLI flow

Sounds like exactly what we’re after in this case! Though I realize it may be tricky to have it work with ‘lamdera format ….’ versus ‘elm-format ….’ I.e the extra command, but have a bash, should be possible.

CharlonTank committed:
Special-case format command to delegate directly to elm-format

As suggested in PR review, the format command now bypasses Lamdera's
normal CLI processing and delegates directly to elm-format's mainIO.
This ensures perfect compatibility with all elm-format features and
reduces code duplication.

Changes:
- Intercept "format" command early in Terminal.hs
- Add minimal FormatDirect module that calls elm-format
- Remove 244-line Format.hs implementation
- Keep stub command for help display consistency

CharlonTank commented:
Hey @supermario ! I've implemented the special-case approach as you suggested. 

The implementation is minimal - just 11 lines in FormatDirect.hs to call elm-format, and a simple conditional check in Terminal.hs.

supermario commented:
@CharlonTank nice one, a few little cleanups I can't seem to attach to this PR, so merging now. Thank you! 🙌
